### ISSUE TITLE
agent surface: serve Ptah's own documentation as an indexed MCP tool

### ISSUE THREAD

**Issue by denisvmedia** [stokaro/ptah#2123]
Serve Ptah's own documentation as an indexed MCP tool, so a model driving the agent surface can answer a question about Ptah from Ptah rather than from training data.

Related: #1483 (the AI integration epic). This is a new operation on the surface ADR 0006 defines, not a change to one.

## The gap

The agent surface can read the project it is pointed at. It knows nothing about **Ptah**.

Measured on `master` at `772bcff0`, a server started with a workspace offers eight tools:

```text
describe_session  validate_schema  render_schema  schema_lineage  read_database
read_artifact     preview_patch    apply_patch
```

None of them can answer "what does `--allow-database-inspect` do", "which dialects support partial indexes", or "what is the difference between `migrations checkpoint` and `migrations baseline`". A model asked those answers from its training data, and for a pre-GA tool that data is either absent or describes a different tool with a similar name.

The failure mode is not a refusal — it is a confident wrong answer, produced by the one component on this surface that nothing verifies.

## What already exists to build on

| | |
| --- | --- |
| site pages | 66 (`docs/site/src/content/docs`, ~1.4 MB) |
| repository docs | 32 (`docs/*.md`, ~0.6 MB) |
| decision records | 8 (`docs/adr`, 116 KB) |
| generated reference | `docs/agent-surface.md`, `docs/capabilities.md`, `docs/agent-errors.md`, the feature matrix |

`npm run build` already produces a search index: `docs/site/dist/pagefind`, 1.7 MB, built by Starlight's pagefind integration over 97 HTML files. Whether that index is the right one to serve is an open question below — it is built for a browser — but it means "index Ptah's documentation" is not starting from nothing.

Nothing is embedded in the binary today. `grep -rn "go:embed"` finds one example and one comment.

## What this is not

- **Not a network fetch.** `network.arbitrary` is hard-denied and no layer may grant it. Schema sources already refuse anything carrying a scheme for exactly this reason (ADR 0006 §2.6). A docs tool that reached docs.ptah.dev would be the same hole with a friendlier name.
- **Not a website scrape**, for the same reason and also because the answer would drift from the binary in the operator's hand.
- **Not a replacement for reading the documentation.** It answers a model's question mid-task; a person still reads the page.

## The questions this needs to answer, not assume

**1. Where do the documents come from at runtime?**

The binary ships without them. Three shapes, and the trade is real:

- *embedded* — `go:embed` an index at build time. Always present, always matches the binary, costs binary size (~2 MB of markdown, less if only an index and passages are kept).
- *operator-named path* — `--docs-dir`, resolved like the schema source roots. No binary cost, and absent for most users, which makes the tool's availability configuration-dependent.
- *separate artifact* — a docs bundle fetched out of band and cached. Most flexible, most machinery, and the fetch has to happen somewhere that is not this process.

**2. Which capability authorizes it?**

ADR 0006's invariant: every operation exposed through the Agent API is authorized by the same effective policy Ptah reports for it. A docs tool needs a capability, and inventing one silently is how the last gap happened.

A new `docs.read` seems right — it reveals nothing about the operator's project or database, so a builtin `allow` is defensible. That is an argument, not a decision: it should be written down in the record, with the reason it is not `project.read` (Ptah's documentation is not the project's content).

**3. What happens when the docs and the binary disagree?**

The site is versioned; `DOCS_VERSION` selects which. An answer from documentation that does not describe the binary being driven is worse than no answer, because it is indistinguishable from a right one. Either the version is bound at build time, or every answer names the version it came from — and `describe_session` reports which is loaded.

**4. Is pagefind the right index?**

It is built for a browser: WASM, chunked fragments, ranked for a search box. Serving it from Go means either linking a WASM runtime or reimplementing the query side. A plain inverted index over 2 MB of markdown is also cheap, and produces passages rather than page hits.

Worth measuring both before choosing, on real questions — "how do I check a migration is reversible", "which flag pins the server version" — rather than on synthetic ones.

**5. What does a call return?**

A model needs a passage; a person verifying it needs a path. Both, then: the matched text, the document it came from as a repository-relative path, and the heading it sits under. Returning a whole page spends context on the 95% that was not asked about.

**6. How does discovery report it?**

`describe_session` separates authority from reachability. "Documentation is loaded, version X, N documents" is reachability. A tool that is offered but has no index behind it is the shape ADR 0006 §2.7 exists to prevent.

## Acceptance

- A model driving `ptah mcp`, with no network, answers a question about Ptah's own behavior and cites the document it came from.
- The answer names the documentation version, and `describe_session` reports the same one.
- The tool is authorized like every other operation: denying its capability stops it, and the denial is what `describe_session` reported.
- The index is built by a gate, so documentation that changed without the index being rebuilt fails CI rather than serving a stale answer.
- Measured: a question whose answer exists in the docs is found; a question whose answer does not exist returns nothing rather than the nearest paragraph.

## Why it is worth doing

Every other tool on this surface exists to stop the model guessing about the project. This one stops it guessing about Ptah — and unlike the others, the guess it replaces is one nobody currently notices, because a wrong answer about a flag looks exactly like a right one until somebody runs it.

**Comment by denisvmedia** [comment]
Taking this. Slicing it so each part lands with its own measurement rather than as one change nobody can review.

**Slice 1 — the index** (#2152, open): a BM25 inverted index over heading-scoped passages. Answers question 4 by building the alternative to pagefind and measuring it: 110 documents, 1385 passages, five real questions each answered, the off-topic control returning nothing. No wiring to the agent surface.

Remaining, in the order I mean to do them:

**Slice 2 — where the documents come from** (question 1). Embedded, on the argument the issue already makes: always present, always matching the binary, and the tool's availability stops being configuration-dependent. The cost is binary size, which I will measure before and after rather than estimate.

**Slice 3 — the capability** (question 2). A new `docs.read` with a builtin `allow`, written into the record with the reason it is not `project.read`: Ptah's documentation is not the project's content, and the tool reveals nothing about the operator's project or database. Denying it must stop the tool, and the denial must be what `describe_session` reported.

**Slice 4 — the tool and its result shape** (question 5). Passage text, repository-relative path, heading. Plus `describe_session` reporting the index as reachability, not authority (question 6).

**Slice 5 — version binding and the freshness gate** (question 3, and the acceptance bullet about CI). An answer from documentation that does not describe the binary being driven is worse than no answer, and documentation that changed without the index being rebuilt has to fail CI rather than serve a stale answer.

One thing I will not carry silently: slice 1's off-topic rule does not reach every question. "how do I reset my forgotten password" still returns one weak hit against a corpus that does discuss passwords. That is in #2152's body with the number, not tuned away.

**Comment by denisvmedia** [comment]
All five slices are open. The acceptance list, measured against the built binary over stdio with no network:

- **A model answers a question about Ptah's own behavior and cites the document** — `docs/site/src/content/docs/versioned/checkpoints.md`, heading `Create a checkpoint > Checkpoint versus baseline`, for "difference between migrations checkpoint and migrations baseline".
- **The answer names the documentation version, and `describe_session` reports the same one** — same version and commit in both, held together by a test.
- **The tool is authorized like every other operation** — with `docs.read deny` in `.ptah/agent-policy`, `describe_session` reported `decided_by: project, verdict: deny` and the call was refused with that same reason.
- **A question whose answer exists is found; one whose answer does not exist returns nothing** — five real questions each answered, the off-topic control empty.

One acceptance bullet is answered differently than it was written. It asks for a gate so documentation that changed without the index being rebuilt fails CI. There is no index to rebuild: the binary carries the markdown, so nothing can be stale against the tree. What that gate would have caught is caught instead by measuring the embed patterns against the repository. Adding a check that cannot fail would read as coverage without being any.

The answers to the six questions:

1. **Where the documents come from** — embedded, as written. An operator-named path makes availability configuration-dependent; a prebuilt index has to be committed, regenerated by every documentation edit, and merges as an opaque blob. Cost measured: 2 203 440 bytes, 2.1% of the binary.
2. **Which capability** — `docs.read`, builtin `allow`, recorded as ADR 0006 §2.8. Not `project.read`, and the separation is measured: denying `project.read` leaves `docs.read` allowed.
3. **Version disagreement** — unexpressible, because the documentation compiles from the same tree. Every answer names the build anyway.
4. **pagefind** — no. It is built for a browser and ranks pages; an answer a model can act on is a passage. A plain BM25 index over 1345 passages was built and measured instead.
5. **What a call returns** — the passage, the document as a repository-relative path, and the heading trail.
6. **How discovery reports it** — a `documentation` section on `describe_session`, as reachability beside the databases and source roots.

Two things I am not carrying silently:

- The off-topic rule does not reach every question. "how do I reset my forgotten password" still returns one weak hit (2.06, against 13.92 for a real answer) against a corpus that does discuss passwords. Tuning a floor until my examples pass would be fitting to the test set.
- `go.5x5.cz/ptah/docs` had to become a public package, because `go:embed` cannot leave its own directory. `scripts/check-public-api.sh` refused it until it was declared with a reason; its whole surface is one variable.

**Comment by denisvmedia** [comment]
All five slices are on master: #2152 (index), #2154 (embedding), #2155 (capability), #2157 (tool), #2158 (version binding).

Verified on a binary built from master, over stdio, with no network:

```text
search_docs "which flag pins the server version"
  documentation: {version: v0.2.1-0.20260825010608-844f259ecdcb,
                  commit:  844f259ecdcbcde6fee225112d0b8a7f35b01faa}
  docs/site/src/content/docs/reference/native-commands.md
    :: Desired schema: `ptah schema` > Live databases: `ptah db`
       > Reading a server's capability profile
       > When configuration, not the version, decides a capability

describe_session
  documentation: {version: …844f259ecdcb, commit: 844f259e…,
                  documents: 108, passages: 1377}
```

The answer names the documentation it came from and `describe_session` reports the same one. Earlier, with `docs.read deny` in `.ptah/agent-policy`, the report showed `decided_by: project, verdict: deny` and the call was refused with that same reason.

One acceptance bullet is answered differently than it was written, and that is deliberate. It asks for a gate so documentation that changed without the index being rebuilt fails CI. There is no index to rebuild — the binary carries the markdown, so nothing can be stale against the tree. What that gate would have caught is caught instead by measuring the embed patterns against the repository. A check that cannot fail would read as coverage without being any.

Two things not carried silently:

* The empty-result rule does not reach every off-topic question. "how do I reset my forgotten password" still returns one weak hit (2.06, against 13.92 for a real answer) against a corpus that does discuss passwords. Tuning a floor until my examples passed would have been fitting to the test set.
* `go.5x5.cz/ptah/docs` had to become a public package, because `go:embed` patterns cannot leave their own directory. `scripts/check-public-api.sh` refused it until it was declared with the reason; its whole surface is one variable.

Closing.
